Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,257 @@
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Address = Message.Sender;
+
+        SetUser(battleId, Message.Sender, user);
+
+        uint userindex = GetUserIndex(battleId);
+        battle.Users.SetValue(user.Address, userindex);
+        SetUserIndex(battleId, (userindex + 1));
+
+        SetBattle(battleId, battle);
+
+        Log(new BattleEventLog { Event = "Enter", BattleId = battleId, Address = Message.Sender });
+    }
+
+    /// <summary>
+    /// 4 different user will end the battle and submit the score
+    /// </summary>
+    public void EndBattle(Address userAddress, ulong battleId, uint score)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        uint ScoreSubmittedCount = GetScoreSubmittedCount(battleId);
+        ScoreSubmittedCount += 1;
+        if (ScoreSubmittedCount == MaxUsers)
+            ProcessWinner(battle);
+
+        SetScoreSubmittedCount(battleId, ScoreSubmittedCount);
+        Log(new BattleEventLog { Event = "End", BattleId = battleId, Address = Message.Sender });
+    }
+
+    /// <summary>
+    /// Get winner address
+    /// </summary>
+    public Address GetWinner(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+        return battle.Winner;
+    }
+
+    /// <summary>
+    /// Process winner when all user scores are submitted
+    /// </summary>
+    private void ProcessWinner(BattleMain battle)
+    {
+        uint winnerIndex = GetWinnerIndex(battle.BattleId, battle.Users);
+        battle.Winner = battle.Users[winnerIndex];
+        SetBattle(battle.BattleId, battle);
+        ProcessPrize(battle.BattleId);
+    }
+
+    /// <summary>
+    /// Get winner user index from battle users
+    /// </summary>
+    private uint GetWinnerIndex(ulong battleid, Address[] users)
+    {
+        uint winningScore = 0;
+        uint winningScoreIndex = 0;
+        for (uint i = 0; i < users.Length; i++)
+        {
+            var user = GetUser(battleid, users[i]);
+            if (user.Score > winningScore)
+            {
+                winningScore = user.Score;
+                winningScoreIndex = i;
+            }
+        }
+        return winningScoreIndex;
+    }
+
+    /// <summary>
+    /// Send 3/4 amount to winner and 1/4 amount to battle owner
+    /// </summary>
+    private void ProcessPrize(ulong battleid)
+    {
+        var battle = GetBattle(battleid);
+        ulong prize = battle.Fee * (MaxUsers - 1);
+        Transfer(battle.Winner, prize);
```

Comment:
If winner address is a contract address then this transfer has possibility to fail. So you can use withdrawal pattern(pull) in here. It will transfer funds immediately if destination is not a contract address and if it is then it will fallback to withdrawal pattern. 

https://github.com/stratisproject/CirrusSmartContracts/blob/cdf12db1467de464ff66516c78e236476124ef3c/Mainnet/NFTStore-Royalties/NFTStore/NFTStore.cs#L68

    private bool SafeTransfer(Address to, ulong amount)
    {
        if (State.IsContract(to))
        {
            var balance = GetBalance(to) + amount;

            SetBalance(to, balance);

            return true;
        }

        var result = Transfer(to, amount);

        return result.Success;
    }

    public bool Withdraw()
    {
        EnsureNotPayable();

        var amount = GetBalance(Message.Sender);

        Assert(amount > 0);

        SetBalance(Message.Sender, 0);

        var transfer = Transfer(Message.Sender, amount);

        Assert(transfer.Success, "Transfer failed.");

        Log(new BalanceRefundedLog { To = Message.Sender, Amount = amount });

        return transfer.Success;
    }

    public ulong GetBalance(Address address)
    {
        return State.GetUInt64($"Balance:{address}");
    }

    private void SetBalance(Address address, ulong balance)
    {
        State.SetUInt64($"Balance:{address}", balance);
    }

Reply:
Winner address will not be a contract address. 4 players will play the game with their wallet address.